File: examples/python/Beta/classification_sequence_parser.py
#!/usr/bin/env python3

import cv2
import depthai as dai


def main() -> None:
    with dai.Pipeline() as pipeline:
        modelSlug = "luxonis/paddle-text-recognition:320x48"
        modelDescription = dai.NNModelDescription(
            modelSlug,
            platform=pipeline.getDefaultDevice().getPlatformAsString(),
        )
        modelArchive = dai.NNArchive(dai.getModelFromZoo(modelDescription))

        cameraNode = pipeline.create(dai.node.Camera).build()
        neuralNetwork = pipeline.create(dai.node.NeuralNetwork).build(cameraNode, modelArchive)
        parserNode = pipeline.create(dai.beta.node.ClassificationSequenceParser).build(
            neuralNetwork.out,
            modelArchive,
        )

        frameQueue = neuralNetwork.passthrough.createOutputQueue()
        outputQueue = parserNode.out.createOutputQueue()
        configQueue = parserNode.inputConfig.createInputQueue()
        config = parserNode.initialConfig

        pipeline.start()
        print("Controls: 't' toggle remove duplicates, 'q' quit.")

        while pipeline.isRunning():
            frameMessage = frameQueue.get()
            parserOutput = outputQueue.get()
            frame = frameMessage.getCvFrame()
            separator = "" if all(len(label) <= 1 for label in parserOutput.classes) else " "
            decodedText = separator.join(parserOutput.classes)
            cv2.putText(
                frame,
                decodedText,
                (20, 35),
                cv2.FONT_HERSHEY_SIMPLEX,
                0.65,
                (0, 255, 0),
                2,
            )


            cv2.imshow("ClassificationSequenceParser", frame)
            key = cv2.waitKey(1)
            if key == ord("q"):
                break
            if key == ord("t"):
                config.removeDuplicates = not config.removeDuplicates
                configQueue.send(config)
                print(f"Remove duplicates: {config.removeDuplicates}")


if __name__ == "__main__":
    main()

File: examples/cpp/Beta/classification_sequence_parser.cpp
#include <algorithm>
#include <cstddef>
#include <iomanip>
#include <iostream>
#include <memory>
#include <opencv2/opencv.hpp>
#include <sstream>
#include <string>

#include "depthai/beta/datatypes.hpp"
#include "depthai/beta/nodes.hpp"
#include "depthai/depthai.hpp"

int main() {
    dai::Pipeline pipeline;

    const std::string modelSlug = "luxonis/paddle-text-recognition:320x48";
    dai::NNModelDescription modelDescription{modelSlug};
    modelDescription.platform = pipeline.getDefaultDevice()->getPlatformAsString();
    dai::NNArchive modelArchive(dai::getModelFromZoo(modelDescription));

    auto cameraNode = pipeline.create<dai::node::Camera>()->build();
    auto neuralNetwork = pipeline.create<dai::node::NeuralNetwork>()->build(cameraNode, modelArchive);
    auto parserNode = pipeline.create<dai::beta::node::ClassificationSequenceParser>()->build(neuralNetwork->out, modelArchive);

    auto frameQueue = neuralNetwork->passthrough.createOutputQueue();
    auto outputQueue = parserNode->out.createOutputQueue();
    auto configQueue = parserNode->inputConfig.createInputQueue();
    auto config = parserNode->initialConfig;

    pipeline.start();
    std::cout << "Controls: 't' toggle remove duplicates, 'q' quit." << std::endl;

    while(pipeline.isRunning()) {
        auto frameMessage = frameQueue->get<dai::ImgFrame>();
        auto parserOutput = outputQueue->get<dai::beta::Classifications>();
        cv::Mat frame = frameMessage->getCvFrame();
        const bool characterSequence =
            std::all_of(parserOutput->classes.begin(), parserOutput->classes.end(), [](const auto& label) { return label.size() <= 1; });
        std::ostringstream decodedText;
        for(std::size_t index = 0; index < parserOutput->classes.size(); ++index) {
            if(index > 0 && !characterSequence) decodedText << ' ';
            decodedText << parserOutput->classes[index];
        }
        cv::putText(frame, decodedText.str(), cv::Point(20, 35), cv::FONT_HERSHEY_SIMPLEX, 0.65, cv::Scalar(0, 255, 0), 2);

        cv::imshow("ClassificationSequenceParser", frame);
        const int key = cv::waitKey(1);
        if(key == 'q') break;
        if(key == 't') {
            config->removeDuplicates = !config->removeDuplicates;
            configQueue->send(config);
            std::cout << "Remove duplicates: " << config->removeDuplicates << std::endl;
        }
    }

    return 0;
}
